calour/io.py

```diff
@@ -302,10 +304,21 @@ def read(data_file, sample_metadata_file=None, feature_metadata_file=None,
     else:
         raise ValueError('unkown data_file_type %s' % data_file_type)
 
+    sid = [str(x) for x in sid]
+    # if we need to cut each table sampleID at the given separator, do it
+    if cut_sample_id_sep is not None:
+        logger.debug('cutting sampleIDs at %s' % cut_sample_id_sep)
+        # sid = sid.map(lambda x: x.split(cut_sample_id_sep)[0])
+        sid = [x.split(cut_sample_id_sep)[0] for x in sid]
+
     sample_metadata = _read_metadata(sid, sample_metadata_file, sample_metadata_kwargs)
     feature_metadata = _read_metadata(fid, feature_metadata_file, feature_metadata_kwargs)
 
-    # store the abund  per sample/feature before any procesing
+    # store the sample and feature ids also as a column (for sorting, etc.)
+    sample_metadata['_sample_id'] = sample_metadata.index.values
+    feature_metadata['_feature_id'] = feature_metadata.index.values
```

Comment:
we wanna minimize the changes in the read(). As `cut_sample_id_sep`, I don't think we should do this.

Reply:
I agree in general we should try to minimize the read() changes.
While we could write the read_ms() without these new fields, they are useful for regular analysis (when you want to sort your samples by their id). So i think while this change is not related to read_ms, it is more elegant this way (where we can treat the featureid and sampleid as regular metadata fields as well). And this does not affect any backwards compatibility (just adds 2 new fields in the exp metadata), so should not break anything.

Reply:
Reasons:

1. I agree it is needed regularly for MS data thus this parameter and the lines of code should be put in read_ms() instead of read(). I have come across any other usage for this to read other data set.

2. this ID manipulation is very specific to MS data - splitting and getting 1st component. so should be put in read_ms()

Reply:
And if we use it in `read_ms`, rename it to `sample_id_splitter` or `sample_id_sep`?

Reply:
there are 2 things which i think are getting confused:
1. the "cut_sample_id_sep" variable - while i agree it is useful only in read_ms, i don't see a way we can move it from read. This because otherwise sampleIDs don't match between mapping file and data table, and read() will fail when it tries to link the two (in _read_metadata() line 222 - set_index().
While i agree users should just have their mapping file and table sampleIDs the same, this is not the common practice in MS, and i don't think we should re-educate all MS users (we will just lose them...)

2. the "_sample_id" and "_feature_id" fields which i added to the sample_metadata and feature_metadata respectively. This is not critical and can be removed. However, currently we don't have these values as a column - just the index, and this means we cannot sort/filter/etc. by them. While usually there is another field to sort/filter by, i encountered some cases where i wanted to sort by the sampleID and this is why i added it. Can be removed if you insist but i think it's elegant.

Reply:
fixed. changed cut_sample_id_sep to callable (and also one for feature_id cutting).
we keep the _sample_id and _feature_id fields since they don't bother and may be useful :)